apps/coordinator/src/actions/transactionActions.js

```diff
@@ -300,97 +309,179 @@ export function setMaxSpendOnOutput(outputIndex) {
-            multisig: slice.multisig,
-            bip32Path: slice.bip32Path,
-            change: slice.change,
-          };
-          inputs.push(input);
-        }
-      });
-    });
+      if (psbt.txInputs.length === 0) {
+        throw new Error("PSBT does not contain any inputs.");
+      }
+      if (psbt.txOutputs.length === 0) {
+        throw new Error("PSBT does not contain any outputs.");
+      }
 
-    if (inputs.length === 0) {
-      throw new Error("PSBT does not contain any UTXOs from this wallet.");
-    }
-    if (inputs.length !== psbt.txInputs.length) {
-      throw new Error(
-        `Only ${inputs.length} of ${psbt.txInputs.length} PSBT inputs are UTXOs in this wallet.`,
-      );
-    }
+      dispatch(resetOutputs());
+      dispatch(setUnsignedPSBT(psbt.toBase64()));
 
-    dispatch(setInputs(inputs));
+      // ==== PROCESS INPUTS ====
+      const inputs = processInputsFromPSBT(psbt, state);
 
-    let outputsTotalSats = new BigNumber(0);
-    psbt.txOutputs.forEach((output, outputIndex) => {
-      const number = outputIndex + 1;
-      outputsTotalSats = outputsTotalSats.plus(BigNumber(output.value));
-      if (number > 1) {
-        dispatch(addOutput());
+      if (inputs.length === 0) {
+        throw new Error("PSBT does not contain any UTXOs from this wallet.");
       }
-
-      if (output.script) {
-        dispatch(setChangeOutputIndex(number));
-        dispatch(setChangeAddressAction(output.address));
+      if (inputs.length !== psbt.txInputs.length) {
+        throw new Error(
+          `Only ${inputs.length} of ${psbt.txInputs.length} PSBT inputs are UTXOs in this wallet.`,
+        );
       }
-      dispatch(setOutputAddress(number, output.address));
-      dispatch(setOutputAmount(number, satoshisToBitcoins(output.value)));
-    });
 
-    state = getState();
-    const inputsTotalSats = BigNumber(state.spend.transaction.inputsTotalSats);
-    const feeSats = inputsTotalSats - outputsTotalSats;
-    const fee = satoshisToBitcoins(feeSats);
-    dispatch(setFee(fee));
+      dispatch(setInputs(inputs));
 
-    dispatch(finalizeOutputs(true));
+      // ==== PROCESS INPUTS ====
+      const { outputsTotalSats } = processOutputsFromPSBT(psbt, dispatch);
 
-    // In the future, if we want to support loading in signatures
-    // (or sets of signatures) included in a PSBT, we likely need to do
-    // that work here. Initial implementation just ignores any signatures
-    // included with the uploaded PSBT.
+      // Calculate and set fee
+      state = getState(); //Now we get updated state after setting inputs
+      const inputsTotalSats = BigNumber(
+        state.spend.transaction.inputsTotalSats,
+      );
+      const feeSats = inputsTotalSats.minus(outputsTotalSats);
+      const fee = satoshisToBitcoins(feeSats);
+      dispatch(setFee(fee));
+
+      // ==== Extract and import signatures ====
+      try {
+        const signatureSets = extractSignaturesFromPSBT(psbt, inputs);
```

Comment:
if you turn processInputs into a selector, then this can basically be another selector as well that can be composed with it because `inputs` would come from the state given a psbt. so `extractSignaturesFromPsbt` could only require the psbt as an argument, inputs would come from the other selector (to be built), `processInputs` or new name like `selectSpendableInputsFromPsbt` (or something) which can be used within this selector. This makes sure to keep the state management clean, and leverages memoization.

Reply:
In fact, I bet this whole bit that's nested in a try/catch could just be put into its own action creator, which itself would call the selectors for what it needs. 

So then, all you would have to do is in this line would be `dispatch(setSignaturesFromPsbt(psbt))`, that itself would be a thunk-based action creator which could take the psbt and the state and then do what you're doing here (but more of it consolidated into selectors)

Reply:
Yep!